# Python Programming: Naive Bayes - Exploratory Data Analysis

## 1. Defining the Question

### a) Specifying the Data Analytic Question

Predict whether an e-mail is a spam or not.

### b) Defining the Metric for Success

A Naive Bayes model thats at least 80% accurate.

### c) Recording the Experimental Design

1. Loading, reading and checking the data. 
2. Data cleaning
3. Exploratory Analysis
4. Modelling and implementing our solution
5. Challenging the solution.

### d) Data Relevance

The data was collected from research on a collection of e-mails. It is therefore relevant to our study.

## 2. Reading the Data

In [3]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import cross_val_score
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.metrics import confusion_matrix,accuracy_score
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.metrics import accuracy_score
from sklearn.naive_bayes import GaussianNB

In [4]:
# load the data
#
spam = pd.read_csv('spambase.data', delimiter=',')
spam



0  0.64  0.64.1  0.1  0.32   0.2   0.3   0.4   0.5   0.6  ...   0.40  \
0     0.21  0.28    0.50  0.0  0.14  0.28  0.21  0.07  0.00  0.94  ...  0.000   
1     0.06  0.00    0.71  0.0  1.23  0.19  0.19  0.12  0.64  0.25  ...  0.010   
2     0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.000   
3     0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.000   
4     0.00  0.00    0.00  0.0  1.85  0.00  0.00  1.85  0.00  0.00  ...  0.000   
...    ...   ...     ...  ...   ...   ...   ...   ...   ...   ...  ...    ...   
4595  0.31  0.00    0.62  0.0  0.00  0.31  0.00  0.00  0.00  0.00  ...  0.000   
4596  0.00  0.00    0.00  0.0  0.00  0.00  0.00  0.00  0.00  0.00  ...  0.000   
4597  0.30  0.00    0.30  0.0  0.00  0.00  0.00  0.00  0.00  0.00  ...  0.102   
4598  0.96  0.00    0.00  0.0  0.32  0.00  0.00  0.00  0.00  0.00  ...  0.000   
4599  0.00  0.00    0.65  0.0  0.00  0.00  0.00  0.00  0.00  0.00  ...  0.000   

       0.41  0.42  0.778   0.43   0.44  3.756   61   278  1  
0     0.132   0.0  0.372  0.180  0.048  5.114  101  1028  1  
1     0.143   0.0  0.276  0.184  0.010  9.821  485  2259  1  
2     0.137   0.0  0.137  0.000  0.000  3.537   40   191  1  
3     0.135   0.0  0.135  0.000  0.000  3.537   40   191  1  
4     0.223   0.0  0.000  0.000  0.000  3.000   15    54  1  
...     ...   ...    ...    ...    ...    ...  ...   ... ..  
4595  0.232   0.0  0.000  0.000  0.000  1.142    3    88  0  
4596  0.000   0.0  0.353  0.000  0.000  1.555    4    14  0  
4597  0.718   0.0  0.000  0.000  0.000  1.404    6   118  0  
4598  0.057   0.0  0.000  0.000  0.000  1.147    5    78  0  
4599  0.000   0.0  0.125  0.000  0.000  1.250    5    40  0  

[4600 rows x 58 columns]

In [5]:
# data description
#
with open("spambase.names") as f:
  print(f.read())


| SPAM E-MAIL DATABASE ATTRIBUTES (in .names format)
|
| 48 continuous real [0,100] attributes of type word_freq_WORD 
| = percentage of words in the e-mail that match WORD,
| i.e. 100 * (number of times the WORD appears in the e-mail) / 
| total number of words in e-mail.  A "word" in this case is any 
| string of alphanumeric characters bounded by non-alphanumeric 
| characters or end-of-string.
|
| 6 continuous real [0,100] attributes of type char_freq_CHAR
| = percentage of characters in the e-mail that match CHAR,
| i.e. 100 * (number of CHAR occurences) / total characters in e-mail
|
| 1 continuous real [1,...] attribute of type capital_run_length_average
| = average length of uninterrupted sequences of capital letters
|
| 1 continuous integer [1,...] attribute of type capital_run_length_longest
| = length of longest uninterrupted sequence of capital letters
|
| 1 continuous integer [1,...] attribute of type capital_run_length_total
| = sum of length of uninterrupted sequences of



## 3. Checking the Data

In [6]:
# Determining the no. of records in our dataset
#
spam.count()

0         4600
0.64      4600
0.64.1    4600
0.1       4600
0.32      4600
0.2       4600
0.3       4600
0.4       4600
0.5       4600
0.6       4600
0.7       4600
0.64.2    4600
0.8       4600
0.9       4600
0.10      4600
0.32.1    4600
0.11      4600
1.29      4600
1.93      4600
0.12      4600
0.96      4600
0.13      4600
0.14      4600
0.15      4600
0.16      4600
0.17      4600
0.18      4600
0.19      4600
0.20      4600
0.21      4600
0.22      4600
0.23      4600
0.24      4600
0.25      4600
0.26      4600
0.27      4600
0.28      4600
0.29      4600
0.30      4600
0.31      4600
0.32.2    4600
0.33      4600
0.34      4600
0.35      4600
0.36      4600
0.37      4600
0.38      4600
0.39      4600
0.40      4600
0.41      4600
0.42      4600
0.778     4600
0.43      4600
0.44      4600
3.756     4600
61        4600
278       4600
1         4600
dtype: int64

In [7]:
# Previewing the top of our dataset
#
spam.head()

0  0.64  0.64.1  0.1  0.32   0.2   0.3   0.4   0.5   0.6  ...  0.40  \
0  0.21  0.28    0.50  0.0  0.14  0.28  0.21  0.07  0.00  0.94  ...  0.00   
1  0.06  0.00    0.71  0.0  1.23  0.19  0.19  0.12  0.64  0.25  ...  0.01   
2  0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
3  0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
4  0.00  0.00    0.00  0.0  1.85  0.00  0.00  1.85  0.00  0.00  ...  0.00   

    0.41  0.42  0.778   0.43   0.44  3.756   61   278  1  
0  0.132   0.0  0.372  0.180  0.048  5.114  101  1028  1  
1  0.143   0.0  0.276  0.184  0.010  9.821  485  2259  1  
2  0.137   0.0  0.137  0.000  0.000  3.537   40   191  1  
3  0.135   0.0  0.135  0.000  0.000  3.537   40   191  1  
4  0.223   0.0  0.000  0.000  0.000  3.000   15    54  1  

[5 rows x 58 columns]

In [8]:
# Previewing the bottom of our dataset
#
spam.tail()

0  0.64  0.64.1  0.1  0.32   0.2  0.3  0.4  0.5  0.6  ...   0.40  \
4595  0.31   0.0    0.62  0.0  0.00  0.31  0.0  0.0  0.0  0.0  ...  0.000   
4596  0.00   0.0    0.00  0.0  0.00  0.00  0.0  0.0  0.0  0.0  ...  0.000   
4597  0.30   0.0    0.30  0.0  0.00  0.00  0.0  0.0  0.0  0.0  ...  0.102   
4598  0.96   0.0    0.00  0.0  0.32  0.00  0.0  0.0  0.0  0.0  ...  0.000   
4599  0.00   0.0    0.65  0.0  0.00  0.00  0.0  0.0  0.0  0.0  ...  0.000   

       0.41  0.42  0.778  0.43  0.44  3.756  61  278  1  
4595  0.232   0.0  0.000   0.0   0.0  1.142   3   88  0  
4596  0.000   0.0  0.353   0.0   0.0  1.555   4   14  0  
4597  0.718   0.0  0.000   0.0   0.0  1.404   6  118  0  
4598  0.057   0.0  0.000   0.0   0.0  1.147   5   78  0  
4599  0.000   0.0  0.125   0.0   0.0  1.250   5   40  0  

[5 rows x 58 columns]

In [9]:
# Checking whether each column has an appropriate datatype
#
spam.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4600 entries, 0 to 4599
Data columns (total 58 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       4600 non-null   float64
 1   0.64    4600 non-null   float64
 2   0.64.1  4600 non-null   float64
 3   0.1     4600 non-null   float64
 4   0.32    4600 non-null   float64
 5   0.2     4600 non-null   float64
 6   0.3     4600 non-null   float64
 7   0.4     4600 non-null   float64
 8   0.5     4600 non-null   float64
 9   0.6     4600 non-null   float64
 10  0.7     4600 non-null   float64
 11  0.64.2  4600 non-null   float64
 12  0.8     4600 non-null   float64
 13  0.9     4600 non-null   float64
 14  0.10    4600 non-null   float64
 15  0.32.1  4600 non-null   float64
 16  0.11    4600 non-null   float64
 17  1.29    4600 non-null   float64
 18  1.93    4600 non-null   float64
 19  0.12    4600 non-null   float64
 20  0.96    4600 non-null   float64
 21  0.13    4600 non-null   float64
 22  

## 4. External Data Source Validation

### a.Validation

The data source is valid

## 5. Tidying the Dataset

In [10]:
# Checking for Outliers
#
Q1 = spam.quantile(0.25)
Q3 = spam.quantile(0.75)
IQR = Q3 - Q1
print(IQR)

print((spam < (Q1 - 1.5 * IQR)) | (spam > (Q3 + 1.5 * IQR)))



0           0.00000
0.64        0.00000
0.64.1      0.42000
0.1         0.00000
0.32        0.38250
0.2         0.00000
0.3         0.00000
0.4         0.00000
0.5         0.00000
0.6         0.16000
0.7         0.00000
0.64.2      0.80000
0.8         0.00000
0.9         0.00000
0.10        0.00000
0.32.1      0.10000
0.11        0.00000
1.29        0.00000
1.93        2.64000
0.12        0.00000
0.96        1.27000
0.13        0.00000
0.14        0.00000
0.15        0.00000
0.16        0.00000
0.17        0.00000
0.18        0.00000
0.19        0.00000
0.20        0.00000
0.21        0.00000
0.22        0.00000
0.23        0.00000
0.24        0.00000
0.25        0.00000
0.26        0.00000
0.27        0.00000
0.28        0.00000
0.29        0.00000
0.30        0.00000
0.31        0.00000
0.32.2      0.00000
0.33        0.00000
0.34        0.00000
0.35        0.00000
0.36        0.11000
0.37        0.00000
0.38        0.00000
0.39        0.00000
0.40        0.00000
0.41        0.18800


In [11]:
# Checking for duplicates
#
print(spam.duplicated().sum())


391


In [12]:
# dropping duplicates
#
spam.drop_duplicates(subset = None, keep = 'first', inplace = True)

In [14]:
# rename columns
#
dict = {'0': 'freq_make',
        '0.64': 'freq_address',
        '0.64.1': 'freq_all',
        '0.1': 'fred_3d',
        '0.32': 'freq_our',
        '0.2': 'freq_over',
        '0.3': 'freq_remove',
        '0.4': 'freq_internet',
        '0.5': 'freq_order',
        '0.6': 'freq_mail',
        '0.7': 'freq_receive',
        '0.64.2': 'freq_will',
        '0.8': 'freq_people',
        '0.9': 'freq_report',
        '0.10': 'freq_addresses',
        '0.32.1': 'freq_free',
        '0.11': 'freq_business',
        '1.29': 'freq_email',
        '1.93': 'freq_you',
        '0.12': 'freq_credit',
        '0.96': 'freq_your',
        '0.13': 'freq_font',
        '0.14': 'freq_000',
        '0.15': 'freq_money',
        '0.16': 'freq_hp',
        '0.17': 'freq_hpl',
        '0.18': 'freq_george',
        '0.19': 'freq_650',
        '0.20': 'freq_lab',
        '0.21': 'freq_labs',
        '0.22': 'freq_telnet',
        '0.23': 'freq_857',
        '0.24': 'freq_data',
        '0.25': 'freq_415',
        '0.26': 'freq_85',
        '0.27': 'freq_technology',
        '0.28': 'freq_1999',
        '0.29': 'freq_parts',
        '0.30': 'freq_pm',
        '0.31': 'freq_direct',
        '0.32.2': 'freq_cs',
        '0.33': 'freq_meeting',
        '0.34': 'freq_original',
        '0.35': 'freq_project',
        '0.36': 'freq_re',
        '0.37': 'freq_edu',
        '0.38': 'freq_table',
        '0.39': 'freq_conference',
        '0.40': 'freq_;',
        '0.41': 'freq_(',
        '0.42': 'freq_[',
        '0.778': 'freq_!',
        '0.43': 'freq_$',
        '0.44': 'freq_#',
        '3.756': 'cap_len_av',
        '61': 'cap_len_lon',
        '278': 'cap_len_tot',
        '1': 'spam'}

spam.rename(columns=dict, 
            inplace=True)
spam.head()

freq_make  freq_address  freq_all  fred_3d  freq_our  freq_over  \
0       0.21          0.28      0.50      0.0      0.14       0.28   
1       0.06          0.00      0.71      0.0      1.23       0.19   
2       0.00          0.00      0.00      0.0      0.63       0.00   
3       0.00          0.00      0.00      0.0      0.63       0.00   
4       0.00          0.00      0.00      0.0      1.85       0.00   

   freq_remove  freq_internet  freq_order  freq_mail  ...  freq_;  freq_(  \
0         0.21           0.07        0.00       0.94  ...    0.00   0.132   
1         0.19           0.12        0.64       0.25  ...    0.01   0.143   
2         0.31           0.63        0.31       0.63  ...    0.00   0.137   
3         0.31           0.63        0.31       0.63  ...    0.00   0.135   
4         0.00           1.85        0.00       0.00  ...    0.00   0.223   

   freq_[  freq_!  freq_$  freq_#  cap_len_av  cap_len_lon  cap_len_tot  spam  
0     0.0   0.372   0.180   0.048       5.114          101         1028     1  
1     0.0   0.276   0.184   0.010       9.821          485         2259     1  
2     0.0   0.137   0.000   0.000       3.537           40          191     1  
3     0.0   0.135   0.000   0.000       3.537           40          191     1  
4     0.0   0.000   0.000   0.000       3.000           15           54     1  

[5 rows x 58 columns]

## 6. Exploratory Analysis

In [15]:
# univariate summary
#
spam.describe()

freq_make  freq_address     freq_all      fred_3d     freq_our  \
count  4209.000000   4209.000000  4209.000000  4209.000000  4209.000000   
mean      0.104391      0.112530     0.291390     0.063093     0.325322   
std       0.300036      0.454241     0.515752     1.352647     0.687887   
min       0.000000      0.000000     0.000000     0.000000     0.000000   
25%       0.000000      0.000000     0.000000     0.000000     0.000000   
50%       0.000000      0.000000     0.000000     0.000000     0.000000   
75%       0.000000      0.000000     0.440000     0.000000     0.410000   
max       4.540000     14.280000     5.100000    42.810000    10.000000   

         freq_over  freq_remove  freq_internet   freq_order    freq_mail  ...  \
count  4209.000000  4209.000000    4209.000000  4209.000000  4209.000000  ...   
mean      0.096679     0.117503       0.108026     0.091882     0.248479  ...   
std       0.276059     0.397327       0.410328     0.282174     0.656705  ...   
min       0.000000     0.000000       0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000       0.000000     0.000000     0.000000  ...   
50%       0.000000     0.000000       0.000000     0.000000     0.000000  ...   
75%       0.000000     0.000000       0.000000     0.000000     0.190000  ...   
max       5.880000     7.270000      11.110000     5.260000    18.180000  ...   

            freq_;       freq_(       freq_[       freq_!       freq_$  \
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000   
mean      0.040413     0.144082     0.017380     0.281018     0.076075   
std       0.252562     0.274280     0.105743     0.843387     0.239734   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.073000     0.000000     0.016000     0.000000   
75%       0.000000     0.194000     0.000000     0.331000     0.053000   
max       4.385000     9.752000     4.081000    32.478000     6.003000   

            freq_#   cap_len_av  cap_len_lon  cap_len_tot         spam  
count  4209.000000  4209.000000  4209.000000   4209.00000  4209.000000  
mean      0.045809     5.384282    52.137800    291.18508     0.398670  
std       0.435976    33.151287   199.605834    618.72831     0.489683  
min       0.000000     1.000000     1.000000      1.00000     0.000000  
25%       0.000000     1.627000     7.000000     40.00000     0.000000  
50%       0.000000     2.297000    15.000000    101.00000     0.000000  
75%       0.000000     3.706000    44.000000    273.00000     1.000000  
max      19.829000  1102.500000  9989.000000  15841.00000     1.000000  

[8 rows x 58 columns]

In [16]:
print(spam.median())
print('*' * 50)
print(spam.mode())
print('*' * 50)
print(spam.skew())
print('*' * 50)
print(spam.kurt())

freq_make            0.000
freq_address         0.000
freq_all             0.000
fred_3d              0.000
freq_our             0.000
freq_over            0.000
freq_remove          0.000
freq_internet        0.000
freq_order           0.000
freq_mail            0.000
freq_receive         0.000
freq_will            0.190
freq_people          0.000
freq_report          0.000
freq_addresses       0.000
freq_free            0.000
freq_business        0.000
freq_email           0.000
freq_you             1.360
freq_credit          0.000
freq_your            0.290
freq_font            0.000
freq_000             0.000
freq_money           0.000
freq_hp              0.000
freq_hpl             0.000
freq_george          0.000
freq_650             0.000
freq_lab             0.000
freq_labs            0.000
freq_telnet          0.000
freq_857             0.000
freq_data            0.000
freq_415             0.000
freq_85              0.000
freq_technology      0.000
freq_1999            0.000
f

In [17]:
# Ploting the bivariate summaries and recording our observations
#
spam.corr()

freq_make  freq_address  freq_all   fred_3d  freq_our  \
freq_make         1.000000      0.034114  0.063267  0.005384  0.021740   
freq_address      0.034114      1.000000  0.027609 -0.008936  0.036983   
freq_all          0.063267      0.027609  1.000000 -0.019895  0.065681   
fred_3d           0.005384     -0.008936 -0.019895  1.000000  0.000256   
freq_our          0.021740      0.036983  0.065681  0.000256  1.000000   
freq_over         0.055040      0.005623  0.066608 -0.009167  0.041392   
freq_remove       0.011037      0.077468  0.028630  0.014286  0.135958   
freq_internet    -0.004392      0.013781  0.007263  0.003798  0.023410   
freq_order        0.103818      0.053525  0.077618 -0.001044  0.014482   
freq_mail         0.040799      0.184441  0.025201 -0.003849  0.027232   
freq_receive      0.063007      0.060819  0.055700 -0.013195  0.079165   
freq_will         0.075919      0.026228  0.065723 -0.018576  0.055008   
freq_people       0.061960      0.015628  0.036484 -0.013156  0.020656   
freq_report       0.032249      0.012465  0.004928  0.003377  0.001586   
freq_addresses    0.022439      0.061887  0.104609  0.005373  0.048309   
freq_free         0.065451      0.046615  0.061927  0.005799  0.079199   
freq_business     0.081447      0.022112  0.029889 -0.002831  0.137869   
freq_email        0.051765      0.178408  0.110528  0.016592  0.050814   
freq_you          0.120575      0.055852  0.126490 -0.009932  0.083092   
freq_credit       0.020910     -0.006690  0.030903 -0.005008  0.027838   
freq_your         0.121471      0.117780  0.147883  0.004578  0.142992   
freq_font        -0.025585      0.001249 -0.039446  0.030587 -0.022716   
freq_000          0.126457      0.007145  0.106747  0.014712  0.062136   
freq_money        0.205676      0.061350  0.031186  0.025849  0.000320   
freq_hp          -0.075085     -0.053488 -0.092635 -0.015706 -0.076785   
freq_hpl         -0.063494     -0.044550 -0.066418 -0.014033 -0.079809   
freq_george      -0.048104     -0.011172 -0.081269 -0.009712 -0.062483   
freq_650         -0.050075     -0.032939 -0.051435 -0.010632 -0.064151   
freq_lab         -0.041556     -0.026005 -0.057823 -0.007933  0.039627   
freq_labs        -0.054262     -0.029359 -0.030483 -0.010914 -0.053614   
freq_telnet      -0.038913     -0.015895 -0.036585 -0.007656 -0.042054   
freq_857         -0.031647      0.028897 -0.063558 -0.006969 -0.024256   
freq_data        -0.042731     -0.032530 -0.057977 -0.008248 -0.034750   
freq_415         -0.026335      0.026085 -0.063396 -0.006977 -0.024556   
freq_85          -0.046101     -0.025472 -0.048938 -0.005909 -0.051059   
freq_technology  -0.057318     -0.028196 -0.047555 -0.006312 -0.052211   
freq_1999        -0.058801      0.003843 -0.074279 -0.007440 -0.077684   
freq_parts       -0.008632     -0.012693  0.031767 -0.002724  0.133057   
freq_pm          -0.011215     -0.016159 -0.017389 -0.004508 -0.045513   
freq_direct      -0.035484      0.002379 -0.045985 -0.007656 -0.017580   
freq_cs          -0.005291     -0.019317 -0.030803 -0.006076 -0.052325   
freq_meeting     -0.026562     -0.035865 -0.007592 -0.008320  0.107706   
freq_original    -0.025621      0.037745 -0.052705 -0.009387 -0.054583   
freq_project     -0.022991     -0.029024 -0.056865 -0.006115  0.002770   
freq_re          -0.037874      0.020116 -0.056554 -0.013182 -0.049499   
freq_edu         -0.034472     -0.025952 -0.060529 -0.009434 -0.083315   
freq_table       -0.000956     -0.012319  0.027397 -0.003410 -0.028146   
freq_conference  -0.018914     -0.020919 -0.029448 -0.001862 -0.035040   
freq_;           -0.027262      0.014782 -0.036754 -0.000030 -0.035049   
freq_(           -0.019680     -0.028209 -0.024462 -0.011450 -0.054583   
freq_[           -0.034798     -0.019394 -0.035780 -0.007516 -0.027988   
freq_!            0.059236      0.031702  0.097231 -0.003861  0.019205   
freq_$            0.101934      0.044449  0.073000  0.008269  0.040696   
freq_#           -0.0

In [18]:
# corelation heatmap
#
plt.figure(figsize = (25, 25))
sns.heatmap(spam.corr(),annot=True)
plt.show()

## 7. Implementing the Solution

In [20]:
# drop redundant or highly correlated columns
#
X = spam.iloc[:, 0:57].values

corr_matrix = spam.corr().abs()
upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(bool))
to_drop = [column for column in upper.columns if any(upper[column] > 0.95)]

spam1 = spam.drop(spam[to_drop], axis=1)
spam1




freq_make  freq_address  freq_all  fred_3d  freq_our  freq_over  \
0          0.21          0.28      0.50      0.0      0.14       0.28   
1          0.06          0.00      0.71      0.0      1.23       0.19   
2          0.00          0.00      0.00      0.0      0.63       0.00   
3          0.00          0.00      0.00      0.0      0.63       0.00   
4          0.00          0.00      0.00      0.0      1.85       0.00   
...         ...           ...       ...      ...       ...        ...   
4595       0.31          0.00      0.62      0.0      0.00       0.31   
4596       0.00          0.00      0.00      0.0      0.00       0.00   
4597       0.30          0.00      0.30      0.0      0.00       0.00   
4598       0.96          0.00      0.00      0.0      0.32       0.00   
4599       0.00          0.00      0.65      0.0      0.00       0.00   

      freq_remove  freq_internet  freq_order  freq_mail  ...  freq_;  freq_(  \
0            0.21           0.07        0.00       0.94  ...   0.000   0.132   
1            0.19           0.12        0.64       0.25  ...   0.010   0.143   
2            0.31           0.63        0.31       0.63  ...   0.000   0.137   
3            0.31           0.63        0.31       0.63  ...   0.000   0.135   
4            0.00           1.85        0.00       0.00  ...   0.000   0.223   
...           ...            ...         ...        ...  ...     ...     ...   
4595         0.00           0.00        0.00       0.00  ...   0.000   0.232   
4596         0.00           0.00        0.00       0.00  ...   0.000   0.000   
4597         0.00           0.00        0.00       0.00  ...   0.102   0.718   
4598         0.00           0.00        0.00       0.00  ...   0.000   0.057   
4599         0.00           0.00        0.00       0.00  ...   0.000   0.000   

      freq_[  freq_!  freq_$  freq_#  cap_len_av  cap_len_lon  cap_len_tot  \
0        0.0   0.372   0.180   0.048       5.114          101         1028   
1        0.0   0.276   0.184   0.010       9.821          485         2259   
2        0.0   0.137   0.000   0.000       3.537           40          191   
3        0.0   0.135   0.000   0.000       3.537           40          191   
4        0.0   0.000   0.000   0.000       3.000           15           54   
...      ...     ...     ...     ...         ...          ...          ...   
4595     0.0   0.000   0.000   0.000       1.142            3           88   
4596     0.0   0.353   0.000   0.000       1.555            4           14   
4597     0.0   0.000   0.000   0.000       1.404            6          118   
4598     0.0   0.000   0.000   0.000       1.147            5           78   
4599     0.0   0.125   0.000   0.000       1.250            5           40   

      spam  
0        1  
1        1  
2        1  
3        1  
4        1  
...    ...  
4595     0  
4596     0  
4597     0  
4598     0  
4599     0  

[4209 rows x 57 columns]

In [21]:
# normalize data
#
from locale import normalize

from sklearn.preprocessing import Normalizer
normalizer = Normalizer(norm='l2')
normalizer.transform(X)

array([[2.03297105e-04, 2.71062806e-04, 4.84040726e-04, ...,
        4.95076854e-03, 9.77762266e-02, 9.95187732e-01],
       [2.59683978e-05, 0.00000000e+00, 3.07292708e-04, ...,
        4.25059392e-03, 2.09911216e-01, 9.77710178e-01],
       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,
        1.81192865e-02, 2.04911354e-01, 9.78451717e-01],
       ...,
       [2.53812946e-03, 0.00000000e+00, 2.53812946e-03, ...,
        1.18784459e-02, 5.07625892e-02, 9.98330921e-01],
       [1.22759766e-02, 0.00000000e+00, 0.00000000e+00, ...,
        1.46672345e-02, 6.39373781e-02, 9.97423098e-01],
       [0.00000000e+00, 0.00000000e+00, 1.59858723e-02, ...,
        3.07420620e-02, 1.22968248e-01, 9.83745986e-01]])

In [22]:
# defining the variables and splitting into train and test
#
X = spam1.iloc[:, 0:56].values
y = spam1.iloc[:, 56].values

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [23]:
# Training our model
# 
gaussian = GaussianNB()  
model = gaussian.fit(X_train, y_train) 

In [24]:
# Predicting our test predictors
#
predicted = model.predict(X_test)
print(np.mean(predicted == y_test))

0.8171021377672208


In [25]:
# cross validating the model
#
from sklearn.model_selection import cross_val_score
all_accuracies = cross_val_score(estimator=gaussian, X=X_train, y=y_train, cv=5)

print(all_accuracies.mean())

0.8286281806517607


In [26]:
# Model accuracy, predictions and confusion matrix
#
y_pred = gaussian.predict(X_test)

from sklearn.metrics import classification_report, confusion_matrix
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy: 0.8171021377672208
[[353 142]
 [ 12 335]]
              precision    recall  f1-score   support

           0       0.97      0.71      0.82       495
           1       0.70      0.97      0.81       347

    accuracy                           0.82       842
   macro avg       0.83      0.84      0.82       842
weighted avg       0.86      0.82      0.82       842



Training 70% of the data:

In [27]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [28]:
# Training our model
# 
gaussian = GaussianNB()  
model = gaussian.fit(X_train, y_train) 

In [29]:
# Predicting our test predictors
#
predicted = model.predict(X_test)
print(np.mean(predicted == y_test))

0.8266033254156769


In [30]:
# cross validating the model
#
from sklearn.model_selection import cross_val_score
all_accuracies = cross_val_score(estimator=gaussian, X=X_train, y=y_train, cv=5)

print(all_accuracies.mean())

0.8255227187706827


In [31]:
# Model accuracy, predictions and confusion matrix
#
y_pred = gaussian.predict(X_test)

from sklearn.metrics import classification_report, confusion_matrix
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy: 0.8266033254156769
[[538 199]
 [ 20 506]]
              precision    recall  f1-score   support

           0       0.96      0.73      0.83       737
           1       0.72      0.96      0.82       526

    accuracy                           0.83      1263
   macro avg       0.84      0.85      0.83      1263
weighted avg       0.86      0.83      0.83      1263



Training 60% of the data:

In [32]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=0)

sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [33]:
# Training our model
# 
gaussian = GaussianNB()  
model = gaussian.fit(X_train, y_train) 

In [34]:
# Predicting our test predictors
#
predicted = model.predict(X_test)
print(np.mean(predicted == y_test))

0.831353919239905


In [35]:
# cross validating the model
#
from sklearn.model_selection import cross_val_score
all_accuracies = cross_val_score(estimator=gaussian, X=X_train, y=y_train, cv=5)

print(all_accuracies.mean())

0.8332673267326733


In [36]:
# Model accuracy, predictions and confusion matrix
#
y_pred = gaussian.predict(X_test)

from sklearn.metrics import classification_report, confusion_matrix
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy: 0.831353919239905
[[737 257]
 [ 27 663]]
              precision    recall  f1-score   support

           0       0.96      0.74      0.84       994
           1       0.72      0.96      0.82       690

    accuracy                           0.83      1684
   macro avg       0.84      0.85      0.83      1684
weighted avg       0.86      0.83      0.83      1684



All the trained models have achieved the metric required with the 60% trained data having the heighest accuracy at 83%.

## 8. Challenging the solution

Further analysis and optimization techniques are required to confirm the validity of the accuracy scores achieved.

## 9. Follow up questions

### a). Did we have the right data?

Yes

### b). Do we need other data to answer our question?

More upto date data is required since the data at hand is from 1999.

### c). Did we have the right question?

Yes.